File: suc/swift/020.swift
import SwiftUI

// MARK: - AppTokens
struct AppTokens {
    struct Colors {
        static let primary = Color(red: 255/255, green: 122/255, blue: 0/255)
        static let secondary = Color(red: 255/255, green: 195/255, blue: 0/255)
        static let tertiary = Color(red: 255/255, green: 243/255, blue: 192/255)
        static let background = Color(red: 255/255, green: 248/255, blue: 231/255)
        static let surface = Color(red: 255/255, green: 255/255, blue: 255/255)
        static let surfaceVariant = Color(red: 255/255, green: 235/255, blue: 200/255)
        static let outline = Color(red: 188/255, green: 167/255, blue: 137/255)
        static let success = Color(red: 40/255, green: 167/255, blue: 69/255)
        static let warning = Color(red: 255/255, green: 193/255, blue: 7/255)
        static let error = Color(red: 220/255, green: 53/255, blue: 69/255)
        static let onPrimary = Color(red: 255/255, green: 255/255, blue: 255/255)
        static let onSecondary = Color(red: 30/255, green: 30/255, blue: 30/255)
        static let onTertiary = Color(red: 30/255, green: 30/255, blue: 30/255)
        static let onBackground = Color(red: 30/255, green: 30/255, blue: 30/255)
        static let onSurface = Color(red: 30/255, green: 30/255, blue: 30/255)
    }

    struct TypographyTokens {
        static let display = Font.system(size: 28, weight: .bold)
        static let title = Font.system(size: 18, weight: .medium)
        static let body = Font.system(size: 14, weight: .regular)
        static let label = Font.system(size: 12, weight: .medium)
    }

    struct Shapes {
        static let small = RoundedRectangle(cornerRadius: 8)
        static let medium = RoundedRectangle(cornerRadius: 14)
        static let large = RoundedRectangle(cornerRadius: 24)
    }

    struct Spacing {
        static let xs: CGFloat = 4
        static let sm: CGFloat = 8
        static let md: CGFloat = 12
        static let lg: CGFloat = 16
        static let xl: CGFloat = 24
        static let xxl: CGFloat = 32
    }

    struct ShadowSpec {
        let elevation: CGFloat
        let radius: CGFloat
        let dy: CGFloat
        let opacity: Double
    }

    struct ElevationMapping {
        static let level1 = ShadowSpec(elevation: 2, radius: 4, dy: 2, opacity: 0.12)
        static let level2 = ShadowSpec(elevation: 4, radius: 8, dy: 4, opacity: 0.16)
    }
}

// MARK: - RootScreen View
struct RootScreen: View {
    @State private var knobPosition: CGPoint = CGPoint(x: 90, y: 140) // 初始位置靠左
    @State private var powerOn: Bool = false

    var body: some View {
        VStack(spacing: 0) {
            // 移除顶部状态栏，增加标题上方的空白
            Color.clear
                .frame(height: 40) // 增加标题上方的空白
            
            // Top Bar
            ZStack {
                AppTokens.Colors.background
                Text("Retro Remote Control")
                    .font(AppTokens.TypographyTokens.display)
                    .foregroundColor(AppTokens.Colors.onSurface)
            }
            .frame(height: 64)
            
            ScrollView {
                VStack(alignment: .center, spacing: AppTokens.Spacing.xl) {
                    // Knob Surface
                    ZStack {
                        AppTokens.Colors.surfaceVariant
                        
                        Canvas { context, size in
                            let center = CGPoint(x: size.width / 2, y: size.height / 2)
                            
                            // Background circle
                            let backgroundPath = Path { path in
                                path.addEllipse(in: CGRect(
                                    x: center.x - 100,
                                    y: center.y - 100,
                                    width: 200,
                                    height: 200
                                ))
                            }
                            context.fill(backgroundPath, with: .color(AppTokens.Colors.secondary.opacity(0.2)))
                            
                            // Knob circle - 直径缩小
                            let knobPath = Path { path in
                                path.addEllipse(in: CGRect(
                                    x: knobPosition.x - 16, // 半径16，直径32
                                    y: knobPosition.y - 16,
                                    width: 32,
                                    height: 32
                                ))
                            }
                            context.fill(knobPath, with: .color(AppTokens.Colors.primary))
                        }
                        .gesture(
                            DragGesture(minimumDistance: 0)
                                .onChanged { value in
                                    let newPosition = value.location
                                    let boundedX = max(16, min(264, newPosition.x))
                                    let boundedY = max(16, min(264, newPosition.y))
                                    knobPosition = CGPoint(x: boundedX, y: boundedY)
                                }
                        )
                    }
                    .frame(width: 280, height: 280)
                    .clipShape(AppTokens.Shapes.large)
                    .shadow(
                        color: Color.black.opacity(AppTokens.ElevationMapping.level1.opacity),
                        radius: AppTokens.ElevationMapping.level1.radius,
                        x: 0,
                        y: AppTokens.ElevationMapping.level1.dy
                    )
                    .padding(AppTokens.Spacing.lg)

                    // Power Status
                    Text(powerOn ? "Power: ON" : "Power: OFF")
                        .font(AppTokens.TypographyTokens.title)
                        .foregroundColor(powerOn ? AppTokens.Colors.success : AppTokens.Colors.error)

                    // Buttons Row - 增加按钮间距
                    HStack(spacing: 60) { // 进一步增加按钮间距
                        Button(action: {
                            powerOn.toggle()
                        }) {
                            Text("On")
                                .font(AppTokens.TypographyTokens.title)
                                .foregroundColor(AppTokens.Colors.onPrimary)
                                .frame(width: 80, height: 80)
                                .background(AppTokens.Colors.primary)
                                .clipShape(Circle())
                        }
                        
                        Button(action: {
                            knobPosition = CGPoint(x: 90, y: 140) // 重置到靠左位置
                        }) {
                            VStack(spacing: 0) {
                                Text("Res")
                                Text("et")
                            }
                            .font(AppTokens.TypographyTokens.title)
                            .foregroundColor(AppTokens.Colors.onSecondary)
                            .frame(width: 80, height: 80)
                            .background(AppTokens.Colors.secondary)
                            .clipShape(Circle())
                        }
                    }

                    // Progress Bar - 整个条都是红色
                    AppTokens.Colors.primary
                        .frame(height: 6)
                        .clipShape(Capsule())
                        .padding(.horizontal, AppTokens.Spacing.lg)
                }
                .padding(AppTokens.Spacing.lg)
            }
            .background(AppTokens.Colors.background)
        }
        .background(AppTokens.Colors.background)
        .ignoresSafeArea()
    }
}

// MARK: - App Entry Point
@main
struct RetroRemoteControlApp: App {
    var body: some Scene {
        WindowGroup {
            RootScreen()
                .statusBar(hidden: true) // 隐藏状态栏
        }
    }
}

File: suc/kotlin/020.kt
package com.example.singlefileui

import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.gestures.detectDragGestures
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat

private const val NAME = "020RemoteControlen"
private const val UI_TYPE = "Media"
private const val STYLE_THEME = "Retro Flat"
private const val LANG = "en"
private const val BASELINE_SIZE = "720x1280"

object AppTokens {
    object Colors {
        val primary = Color(0xFFFF7A00)
        val secondary = Color(0xFFFFC300)
        val tertiary = Color(0xFFFFF3C0)
        val background = Color(0xFFFFF8E7)
        val surface = Color(0xFFFFFFFF)
        val surfaceVariant = Color(0xFFFFEBC8)
        val outline = Color(0xFFBCA789)
        val success = Color(0xFF28A745)
        val warning = Color(0xFFFFC107)
        val error = Color(0xFFDC3545)
        val onPrimary = Color(0xFFFFFFFF)
        val onSecondary = Color(0xFF1E1E1E)
        val onTertiary = Color(0xFF1E1E1E)
        val onBackground = Color(0xFF1E1E1E)
        val onSurface = Color(0xFF1E1E1E)
    }
    object TypographyTokens {
        val display = TextStyle(fontSize = 28.sp, fontWeight = FontWeight.Bold)
        val title = TextStyle(fontSize = 18.sp, fontWeight = FontWeight.Medium)
        val body = TextStyle(fontSize = 14.sp, fontWeight = FontWeight.Normal)
        val label = TextStyle(fontSize = 12.sp, fontWeight = FontWeight.Medium)
    }
    object Shapes {
        val small = RoundedCornerShape(8.dp)
        val medium = RoundedCornerShape(14.dp)
        val large = RoundedCornerShape(24.dp)
    }
    object Spacing {
        val xs = 4.dp
        val sm = 8.dp
        val md = 12.dp
        val lg = 16.dp
        val xl = 24.dp
        val xxl = 32.dp
    }
    data class ShadowSpec(val elevation: Dp, val radius: Dp, val dy: Dp, val opacity: Float)
    object ElevationMapping {
        val level1 = ShadowSpec(2.dp, 4.dp, 2.dp, 0.12f)
        val level2 = ShadowSpec(4.dp, 8.dp, 4.dp, 0.16f)
    }
}

private val AppColorScheme = lightColorScheme(
    primary = AppTokens.Colors.primary,
    onPrimary = AppTokens.Colors.onPrimary,
    secondary = AppTokens.Colors.secondary,
    onSecondary = AppTokens.Colors.onSecondary,
    tertiary = AppTokens.Colors.tertiary,
    onTertiary = AppTokens.Colors.onTertiary,
    background = AppTokens.Colors.background,
    onBackground = AppTokens.Colors.onBackground,
    surface = AppTokens.Colors.surface,
    onSurface = AppTokens.Colors.onSurface,
    surfaceVariant = AppTokens.Colors.surfaceVariant,
    outline = AppTokens.Colors.outline,
    error = AppTokens.Colors.error
)

private val AppTypography = Typography(
    displayLarge = AppTokens.TypographyTokens.display,
    titleMedium = AppTokens.TypographyTokens.title,
    bodyMedium = AppTokens.TypographyTokens.body,
    labelMedium = AppTokens.TypographyTokens.label
)

@Composable
fun AppTheme(content: @Composable () -> Unit) {
    MaterialTheme(
        colorScheme = AppColorScheme,
        typography = AppTypography,
        shapes = Shapes(
            small = AppTokens.Shapes.small,
            medium = AppTokens.Shapes.medium,
            large = AppTokens.Shapes.large
        ),
        content = content
    )
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun RootScreen() {
    var knobPosition by remember { mutableStateOf(Offset(0f, 0f)) }
    var powerOn by remember { mutableStateOf(false) }

    Scaffold(
        contentWindowInsets = WindowInsets(0),
        topBar = {
            CenterAlignedTopAppBar(
                title = { Text("Retro Remote Control", style = MaterialTheme.typography.displayLarge, color = AppTokens.Colors.onSurface) },
                colors = TopAppBarDefaults.centerAlignedTopAppBarColors(containerColor = AppTokens.Colors.background)
            )
        },
        containerColor = AppTokens.Colors.background
    ) { pad ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(pad)
                .background(AppTokens.Colors.background)
                .padding(AppTokens.Spacing.lg),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(AppTokens.Spacing.xl)
        ) {
            Surface(
                modifier = Modifier
                    .size(280.dp)
                    .background(AppTokens.Colors.surfaceVariant, shape = AppTokens.Shapes.large),
                shape = AppTokens.Shapes.large,
                color = AppTokens.Colors.surfaceVariant,
                tonalElevation = AppTokens.ElevationMapping.level1.elevation,
                shadowElevation = AppTokens.ElevationMapping.level1.elevation
            ) {
                Box(
                    modifier = Modifier
                        .fillMaxSize()
                        .padding(AppTokens.Spacing.lg),
                    contentAlignment = Alignment.Center
                ) {
                    Canvas(modifier = Modifier
                        .fillMaxSize()
                        .pointerInput(Unit) {
                            detectDragGestures { change, _ ->
                                knobPosition = change.position
                            }
                        }) {
                        val center = Offset(size.width / 2f, size.height / 2f)
                        drawCircle(
                            color = AppTokens.Colors.secondary.copy(alpha = 0.2f),
                            radius = 200f,
                            center = center
                        )
                        drawCircle(
                            color = AppTokens.Colors.primary,
                            radius = 24f,
                            center = knobPosition
                        )
                    }
                }
            }
            Text(
                text = if (powerOn) "Power: ON" else "Power: OFF",
                style = MaterialTheme.typography.titleMedium,
                color = if (powerOn) AppTokens.Colors.success else AppTokens.Colors.error
            )
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceEvenly
            ) {
                Button(
                    onClick = { powerOn = !powerOn },
                    colors = ButtonDefaults.buttonColors(containerColor = AppTokens.Colors.primary, contentColor = AppTokens.Colors.onPrimary),
                    shape = CircleShape,
                    modifier = Modifier.size(80.dp)
                ) {
                    Text(if (powerOn) "Off" else "On", style = MaterialTheme.typography.titleMedium)
                }
                Button(
                    onClick = { knobPosition = Offset.Zero },
                    colors = ButtonDefaults.buttonColors(containerColor = AppTokens.Colors.secondary, contentColor = AppTokens.Colors.onSecondary),
                    shape = CircleShape,
                    modifier = Modifier.size(80.dp)
                ) {
                    Text("Reset", style = MaterialTheme.typography.titleMedium)
                }
            }
            LinearProgressIndicator(
                progress = { (knobPosition.x + knobPosition.y) % 400 / 400f },
                color = AppTokens.Colors.primary,
                trackColor = AppTokens.Colors.surfaceVariant,
                modifier = Modifier.fillMaxWidth().height(6.dp)
            )
        }
    }


}

class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        WindowCompat.setDecorFitsSystemWindows(window, false)
        val controller = WindowInsetsControllerCompat(window, window.decorView)
        controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
        controller.hide(WindowInsetsCompat.Type.systemBars())
        setContent {
            AppTheme {
                Surface(color = MaterialTheme.colorScheme.background) {
                    RootScreen()
                }
            }
        }
    }
    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        if (hasFocus) {
            val controller = WindowInsetsControllerCompat(window, window.decorView)
            controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
            controller.hide(WindowInsetsCompat.Type.systemBars())
        }
    }
}

@Preview(showBackground = true, backgroundColor = 0xFFFFF8E7)
@Composable
fun PreviewScreen() {
    AppTheme {
        RootScreen()
    }
}